server/etcdserver/txn/txn_test.go

```diff
@@ -94,3 +98,253 @@ func TestWriteTxnPanic(t *testing.T) {
+	if _, err := as.UserGrantRole(&pb.AuthUserGrantRoleRequest{User: "foo", Role: "rw"}); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := as.AuthEnable(); err != nil {
+		t.Fatal(err)
+	}
+
+	tests := []struct {
+		name       string
+		txnRequest *pb.TxnRequest
+		err        error
+	}{
+		{
+			name: "Unauthorize out of range compare",
+			txnRequest: &pb.TxnRequest{
+				Compare: []*pb.Compare{
+					{
+						Key: []byte("boo"),
+					},
+				},
+				Success: []*pb.RequestOp{},
+			},
+			err: auth.ErrPermissionDenied,
+		},
+		{
+			name: "No error for nil request range",
+			txnRequest: &pb.TxnRequest{
+				Success: []*pb.RequestOp{
+					{
+						Request: &pb.RequestOp_RequestRange{
+							RequestRange: nil,
+						},
+					},
+				},
+			},
+			err: nil,
+		},
+		{
+			name: "Authorize request range in range",
+			txnRequest: &pb.TxnRequest{
+				Success: []*pb.RequestOp{
+					{
+						Request: &pb.RequestOp_RequestRange{
+							RequestRange: &pb.RangeRequest{
+								Key:      []byte("foo"),
+								RangeEnd: []byte("zoo"),
+							},
+						},
+					},
+				},
+			},
+			err: nil,
+		},
+		{
+			name: "Unauthorize request range out of range",
+			txnRequest: &pb.TxnRequest{
+				Success: []*pb.RequestOp{
+					{
+						Request: &pb.RequestOp_RequestRange{
+							RequestRange: &pb.RangeRequest{
+								Key:      []byte("boo"),
+								RangeEnd: []byte("zoo"),
+							},
+						},
+					},
+				},
+			},
+			err: auth.ErrPermissionDenied,
+		},
+		{
+			name: "No error for nil request put",
+			txnRequest: &pb.TxnRequest{
+				Success: []*pb.RequestOp{
+					{
+						Request: &pb.RequestOp_RequestPut{
+							RequestPut: nil,
+						},
+					},
+				},
+			},
+			err: nil,
+		},
+		{
+			name: "Authorize request pur in range",
+			txnRequest: &pb.TxnRequest{
+				Success: []*pb.RequestOp{
+					{
+						Request: &pb.RequestOp_RequestPut{
+							RequestPut: &pb.PutRequest{
+								Key: []byte("foo"),
+							},
+						},
+					},
+				},
+			},
+			err: nil,
+		},
+		{
+			name: "Unauthorized request pur in range",
+			txnRequest: &pb.TxnRequest{
+				Success: []*pb.RequestOp{
+					{
+						Request: &pb.RequestOp_RequestPut{
+							RequestPut: &pb.PutRequest{
+								Key: []byte("boo"),
+							},
+						},
+					},
+				},
+			},
+			err: auth.ErrPermissionDenied,
+		},
+		{
+			name: "No error for nil delete range",
+			txnRequest: &pb.TxnRequest{
+				Success: []*pb.RequestOp{
+					{
+						Request: &pb.RequestOp_RequestDeleteRange{
+							RequestDeleteRange: nil,
+						},
+					},
+				},
+			},
+			err: nil,
+		},
+		{
+			name: "Authorize delete range in range compare and rerquest",
+			txnRequest: &pb.TxnRequest{
+				Compare: []*pb.Compare{
+					{
+						Key: []byte("foo"),
+					},
+				},
+				Success: []*pb.RequestOp{
+					{
+						Request: &pb.RequestOp_RequestDeleteRange{
+							RequestDeleteRange: &pb.DeleteRangeRequest{
+								Key:      []byte("foo"),
+								RangeEnd: []byte("zoo"),
+								PrevKv:   true,
```

Comment:
Any reason you test DeleteRangeRequest with Compare? Could you test Compare idependently?

Reply:
I separated it.